pkg/distributor/distributor.go

```diff
@@ -519,49 +514,32 @@ func (d *Distributor) Push(ctx context.Context, req *client.WriteRequest) (*clie
 		}
 	}
 
-	if len(metadataKeys) > 0 {
-		err = ring.DoBatch(ctx, subRing, metadataKeys, func(ingester ring.IngesterDesc, indexes []int) error {
-			metadata := make([]*client.MetricMetadata, 0, len(indexes))
-			for _, i := range indexes {
-				metadata = append(metadata, validatedMetadata[i])
-			}
+	keys := append(seriesKeys, metadataKeys...)
+	initialMetadataIndex := len(seriesKeys)
 
-			localCtx, cancel := context.WithTimeout(context.Background(), d.cfg.RemoteTimeout)
-			defer cancel()
-			localCtx = user.InjectUserID(localCtx, userID)
-
-			if sp := opentracing.SpanFromContext(ctx); sp != nil {
-				localCtx = opentracing.ContextWithSpan(localCtx, sp)
-			}
-
-			return d.sendMetadata(localCtx, ingester, metadata)
-		}, func() {})
-		if err != nil {
-			// Metadata is a best-effort approach so we consider failures non-fatal, log them, and move on.
-			logger := util.WithContext(ctx, util.Logger)
-			level.Error(logger).Log("msg", "Failed to send metadata to ingesters", "err", err)
-		}
-	}
+	err = ring.DoBatch(ctx, subRing, keys, func(ingester ring.IngesterDesc, indexes []int) error {
+		timeseries := make([]client.PreallocTimeseries, 0, len(indexes))
+		metadata := make([]*client.MetricMetadata, 0, len(indexes))
```

Comment:
Is this the most efficient way to allocate these spaces? Given we don't really know how many of the indexes are actually for metadata or samples (w/o changing the ring function - which I tried to avoid) we at least know that both of them are less the total number of indexes.

Reply:
Good shout, done.